### PR TITLE
Redo error ABCI interface

### PR DESCRIPTION
**Pull request by husio**
I have started this as a solution to the hidden panic issue https://github.com/iov-one/weave/issues/363

Down the rabbit hole, I am refactoring how errors are handled. Before finishing I want to ask if what I am proposing makes sense and is acceptable.
Before writing yes/no, please let the proposed refactor ideas sink. It took me several iterations and eureka moments to get here :wink: 

1. The most crucial question is **how do we expose the errors to tendermint?** Errors are exposed to tendermint (almost) only through [two functions in `abci.go`](https://github.com/iov-one/weave/blob/master/abci.go#L108-L139). This means that those two functions decide how an error is presented and no other code matters.

2. An important design decision is whether we want to allow non `error.Error` errors to be exposed via the ABCI API (wrapping an `error.Error` counts as the `error.Error`). I think the answer is no -- only `error.Error` instances (wrapped or not) should be exposed via ABCI API. No other error is exposed via ABCI API.
This is why `ABCIInfo` function expose ~only `error.Error`~ certain errors information.

3. Should a functionality explicitely return `errors.ErrInternal`? I think no. I think `ErrInternal` should be removed from the erros definitions so that it cannot be returned. ABCI code 1 should be restricted to trully internal errors and used in [functions that create ABCI response only](https://github.com/iov-one/weave/blob/master/abci.go#L108-L139). There is no need to be able to use it anywhere else.
_(above change is ~not presented here yet~ implemented now - `ErrInternal` is removed)_

Please have a look at the (incomplete) proposed changes and let me know what you think. We can also have a call if taking through those changes will be a better explanation. 
If I get positive feedback, I will finish this refactoring.

### CODE REVIEW

**Comment by ethanfrey** [comment]
I read those three points and generally agree.

Basically, we never explicitly make ErrInternal... we wrap some stdlib error and that is an "internal error". Similar apporach with catching panics (which will contain a stdlib error).

Fine with me not to expose such over abci and use the error code. (Exception for allowing visibility with Debug flag, so let's update Redact use first, so it is not impossible to debug code).

We only then expose full info for wrapped Error that was explicitly returned, right?

File: abci.go [hunk 1]
```diff
@@ -100,40 +98,32 @@ type TickResult struct {
...
+// When in debug mode always the full error information is returned.
 func DeliverTxError(err error, debug bool) abci.ResponseDeliverTx {
-	clean := errors.Redact(err)
-	tm := errors.Wrap(clean, "cannot deliver tx")
-
-	log := tm.ABCILog()
-	if debug {
-		log = fmt.Sprintf("%v", tm)
+	err = errors.Redact(err, debug)
+	code, log := errors.ABCIInfo(err, debug)
```
Comment (ethanfrey) [inline abci.go]: Nice abstraction here, will look at implementation

File: app/store.go [hunk 2]
```diff
@@ -215,7 +215,8 @@ func (s *StoreApp) Query(reqQuery abci.RequestQuery) (resQuery abci.ResponseQuer
 	path, mod := splitPath(reqQuery.Path)
 	qh := s.queryRouter.Handler(path)
 	if qh == nil {
-		resQuery.Code = errors.ErrNotFound.ABCICode()
+		code, _ := errors.ABCIInfo(errors.ErrNotFound)
```
Comment (ethanfrey) [inline app/store.go]: interesting

File: errors/abci.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+import (
+	"errors"
+	"reflect"
+)
+
+// ABCIInfo returns the ABCI error information as consumed by the tenderemint
+// client.
+// This function provides a full error infromation
+func ABCIInfo(err error, debug bool) (uint32, string) {
```
Comment (ethanfrey) [inline errors/abci.go]: Please document return value (code, log)

File: errors/abci.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+	// under an internal error code and a generic message instead of
+	// detailed error string.
+	internalABCICode = 1
+	internalABCILog  = "internal error"
+)
+
+// abciCode test if given error contains an ABCI code and returns the value of
+// it if available. This function is testing for the causer interface as well
+// and unwraps the error.
+func abciCode(err error) uint32 {
```
Comment (ethanfrey) [inline errors/abci.go]: looks clean

File: errors/abci.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+	return false
+}
+
+// Redact replace all errors that do not initialize with a weave error with a
+// generic internal error instance. This function is supposed to hide
+// implementation details errors and leave only those that weave framework
+// originates.
+//
+// This is a no-operation function when running in debug mode.
+func Redact(err error, debug bool) error {
```
Comment (ethanfrey) [inline errors/abci.go]: Looks good

**Comment by husio** [comment]
> We only then expose full info for wrapped Error that was explicitly returned, right?

Yes(ish). I think this is better explained with all the test cases. Please have a look at two tests in https://github.com/iov-one/weave/blob/3436ed02910b117b000888ccd8756a932aa51dcc/errors/abci_test.go

File: errors/abci.go (hunk 3 above)
Comment (husio) [reply errors/abci.go]: Done in https://github.com/iov-one/weave/pull/399/commits/dbd5b8771d4bef5c83b47d4b198b09fd44a3fda1

File: errors/abci.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+		}
+	}
+}
+
+// errIsNil returns true if value represented by the given error is nil.
+//
+// Most of the time a simple == check is enough. There is a very narrowed
+// spectrum of cases (mostly in tests) where a more sophisticated check is
+// required.
+func errIsNil(err error) bool {
```
Comment (alpe) [inline errors/abci.go]: 👍

File: errors/errors.go [hunk 7]
```diff
@@ -191,36 +165,29 @@ func (kind *Error) Is(err error) bool {
...
-	// this will not fire on wrapping a wrappedError,
-	// but only on wrapping a registered Error, or stdlib error
-	st, ok := err.(stackTracer)
-	if !ok {
-		st = errors.WithStack(err).(stackTracer)
+	// If this error does not carry the stacktrace information yet, attach
+	// one. This should be done only once per error at the lowest frame
+	// possible (most inner wrap).
+	if stackTrace(err) == nil {
+		err = errors.WithStack(err)
```
Comment (alpe) [inline errors/errors.go]: Nice use of pkg/errors

File: errors/abci.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+	// For internal errors hide the original error message and return
+	// generic data.
+	return internalABCICode, internalABCILog
+}
+
+const (
+	// ABCI response use 0 to signal that the processing was successful and
+	// no error is returned.
+	notErrorCode = 0
```
[the remark this answers is not in the document]
Comment (ruseinov) [reply errors/abci.go]: yeah, I'd prefer this too, notErrorCode is stumbling. not blocking though

**Comment by ruseinov** [comment]
looks good to me apart from a minor naming nit. good job! This interface asked for refactoring after we did errors.

File: errors/abci.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (husio) [reply errors/abci.go]: Done in https://github.com/iov-one/weave/pull/399/commits/a2f905921568ba138d1e5bade9a2d30555769e1f

File: errors/abci.go [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+	if debug {
+		return internalABCICode, err.Error()
+	}
+
+	// For internal errors hide the original error message and return
+	// generic data.
+	return internalABCICode, internalABCILog
+}
+
+const (
```
[the remark this answers is not in the document]
Comment (husio) [reply errors/abci.go]: Done in https://github.com/iov-one/weave/pull/399/commits/a2f905921568ba138d1e5bade9a2d30555769e1f

**Comment by husio** [comment]
@alpe @ethanfrey I have addressed all comments, please have a look again.

**Pull request merged**